File: db-async-common/src/main/scala/com/github/mauricio/async/db/util/ChannelWrapper.scala
/*
 * Copyright 2013 Maurício Linhares
 *
 * Maurício Linhares licenses this file to you under the Apache License,
 * version 2.0 (the "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at:
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */

package com.github.mauricio.async.db.util

import com.github.mauricio.async.db.exceptions.UnknownLengthException
import java.nio.charset.Charset
import scala.language.implicitConversions
import io.netty.buffer.ByteBuf

object ChannelWrapper {
  implicit def bufferToWrapper( buffer : ByteBuf ) = new ChannelWrapper(buffer)

  final val MySQL_NULL = 0xfb
  final val log = Log.get[ChannelWrapper]

}

class ChannelWrapper( val buffer : ByteBuf ) extends AnyVal {

  import ChannelWrapper._

  def readFixedString( length : Int, charset : Charset ) : String = {
    val bytes = new Array[Byte](length)
    buffer.readBytes( bytes )
    new String( bytes, charset )
  }

  def readCString( charset : Charset ) = ByteBufferUtils.readCString(buffer, charset)

  def readUntilEOF( charset: Charset ) = ByteBufferUtils.readUntilEOF(buffer, charset)

  def readLengthEncodedString( charset : Charset ) : String = {
    val length = readBinaryLength
    readFixedString(length.asInstanceOf[Int], charset)
  }

  def readBinaryLength : Long = {
    val firstByte = buffer.readUnsignedByte()

    if ( firstByte <= 250 ) {
      firstByte
    } else {
      firstByte match {
        case MySQL_NULL => -1
        case 252 => buffer.readUnsignedShort()
        case 253 => readLongInt
        case 254 => buffer.readLong()
        case _ => throw new UnknownLengthException(firstByte)
      }
    }

  }

  def readLongInt : Int = {
    val first = buffer.readByte()
    val second = buffer.readByte()
    val third = buffer.readByte()

    ( first & 0xff ) | (( second & 0xff ) << 8) | ((third & 0xff) << 16)
  }

  def writeLength( length : Long ) {
    if (length < 251) {
      buffer.writeByte( length.asInstanceOf[Byte])
    } else if (length < 65536L) {
      buffer.writeByte(252)
      buffer.writeShort(length.asInstanceOf[Int])
    } else if (length < 16777216L) {
      buffer.writeByte(253)
      writeLongInt(length.asInstanceOf[Int])
    } else {
      buffer.writeByte(254)
      buffer.writeLong(length)
    }
  }

  def writeLongInt(i : Int) {
    buffer.writeByte( i & 0xff )
    buffer.writeByte( i >>> 8 )
    buffer.writeByte( i >>> 16 )
  }

  def writeLenghtEncodedString( value : String, charset : Charset ) {
    val bytes = value.getBytes(charset)
    writeLength(bytes.length)
    buffer.writeBytes(bytes)
  }

  def writePacketLength( sequence : Int = 0 ) {
    ByteBufferUtils.writePacketLength(buffer, sequence )
  }

  def mysqlReadInt() : Int = {
    val first = buffer.readByte()
    val last = buffer.readByte()

    (first & 0xff) | ((last & 0xff) << 8)
  }


}

File: db-async-common/src/main/java/com/github/jasync/sql/db/util/ChannelWrapper.java
package com.github.jasync.sql.db.util;

import com.github.jasync.sql.db.exceptions.UnknownLengthException;
import io.netty.buffer.ByteBuf;

import java.nio.charset.Charset;

public class ChannelWrapper {
  //  implicit public void bufferToWrapper( buffer :ByteBuf) = new ChannelWrapper(buffer)
//
  private static final short MySQL_NULL = 0xfb;
//  final val log = Log.get[ChannelWrapper]

  public static String readFixedString(ByteBuf buffer, int length, Charset charset) {
    byte[] bytes = new byte[length];
    buffer.readBytes(bytes);
    return new String(bytes, charset);
  }

  public static String readCString(ByteBuf buffer, Charset charset) {
    return ByteBufferUtils.readCString(buffer, charset);
  }

  public static String readUntilEOF(ByteBuf buffer, Charset charset) {
    return ByteBufferUtils.readUntilEOF(buffer, charset);
  }

  public static String readLengthEncodedString(ByteBuf buffer, Charset charset) {
    long length = readBinaryLength(buffer);
    return readFixedString(buffer, (int) length, charset);
  }

  public static long readBinaryLength(ByteBuf buffer) {
    short firstByte = buffer.readUnsignedByte();


    if (firstByte <= 250) {
      return firstByte;
    } else {
      switch (firstByte) {
        case MySQL_NULL:
          return -1;
        case 252:
          return buffer.readUnsignedShort();
        case 253:
          return readLongInt(buffer);
        case 254:
          return buffer.readLong();
        default:
          throw new UnknownLengthException(firstByte);
      }
    }

  }

  public static int readLongInt(ByteBuf buffer) {
    byte first = buffer.readByte();
    byte second = buffer.readByte();
    byte third = buffer.readByte();

    return (first & 0xff) | ((second & 0xff) << 8) | ((third & 0xff) << 16);
  }

  public static void writeLength(ByteBuf buffer, long length) {
    if (length < 251) {
      buffer.writeByte((byte) length);
    } else if (length < 65536L) {
      buffer.writeByte(252);
      buffer.writeShort((int) length);
    } else if (length < 16777216L) {
      buffer.writeByte(253);
      writeLongInt(buffer, (int) length);
    } else {
      buffer.writeByte(254);
      buffer.writeLong(length);
    }
  }

  public static void writeLongInt(ByteBuf buffer, int i) {
    buffer.writeByte(i & 0xff);
    buffer.writeByte(i >>> 8);
    buffer.writeByte(i >>> 16);
  }

  public static void writeLenghtEncodedString(ByteBuf buffer, String value, Charset charset) {
    byte[] bytes = value.getBytes(charset);
    writeLength(buffer, bytes.length);
    buffer.writeBytes(bytes);
  }

  // : Int = 0
  public static void writePacketLength(ByteBuf buffer, int sequence) {
    ByteBufferUtils.writePacketLength(buffer, sequence);
  }

  public static int mysqlReadInt(ByteBuf buffer) {
    byte first = buffer.readByte();
    byte last = buffer.readByte();

    return (first & 0xff) | ((last & 0xff) << 8);
  }

}
